io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplus3em/IoShellyPlus3emImpl.java

```diff
@@ -0,0 +1,249 @@
+@EventTopics({ EdgeEventConstants.TOPIC_CYCLE_EXECUTE_WRITE, EdgeEventConstants.TOPIC_CYCLE_AFTER_PROCESS_IMAGE })
+public class IoShellyPlus3emImpl extends AbstractOpenemsComponent implements IoShellyPlus3em, DigitalOutput,
+		SinglePhaseMeter, ElectricityMeter, OpenemsComponent, TimedataProvider, EventHandler {
+
+	private final CalculateEnergyFromPower calculateProductionEnergy = new CalculateEnergyFromPower(this,
+			ElectricityMeter.ChannelId.ACTIVE_PRODUCTION_ENERGY);
+	private final CalculateEnergyFromPower calculateConsumptionEnergy = new CalculateEnergyFromPower(this,
+			ElectricityMeter.ChannelId.ACTIVE_CONSUMPTION_ENERGY);
+
+	private final Logger log = LoggerFactory.getLogger(IoShellyPlus3emImpl.class);
+	private final BooleanWriteChannel[] digitalOutputChannels;
+
+	private MeterType meterType = null;
+	private SinglePhase phase = null;
+	private String baseUrl;
+
+	private volatile Timedata timedata;
+
+	@Reference()
+	private BridgeHttpFactory httpBridgeFactory;
+	private BridgeHttp httpBridge;
+
+	public IoShellyPlus3emImpl() {
+		super(OpenemsComponent.ChannelId.values(), ElectricityMeter.ChannelId.values(),
+				DigitalOutput.ChannelId.values(), IoShellyPlus3em.ChannelId.values());
+		this.digitalOutputChannels = new BooleanWriteChannel[] { this.channel(IoShellyPlus3em.ChannelId.RELAY) };
+
+		ElectricityMeter.calculateSumActivePowerFromPhases(this);
+	}
+
+	@Activate
+	protected void activate(ComponentContext context, Config config) {
+		super.activate(context, config.id(), config.alias(), config.enabled());
+		this.meterType = config.type();
+		this.baseUrl = "http://" + config.ip();
+		this.httpBridge = this.httpBridgeFactory.get();
+
+		if (this.isEnabled()) {
+			this.httpBridge.subscribeJsonEveryCycle(this.baseUrl + "/status", this::processHttpResult);
+		}
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		this.httpBridgeFactory.unget(this.httpBridge);
+		this.httpBridge = null;
+		super.deactivate();
+	}
+
+	@Override
+	public BooleanWriteChannel[] digitalOutputChannels() {
+		return this.digitalOutputChannels;
+	}
+
+	@Override
+	public String debugLog() {
+		var b = new StringBuilder();
+		var valueOpt = this.getRelayChannel().value().asOptional();
+		b.append(valueOpt.isPresent() ? (valueOpt.get() ? "ON" : "OFF") : "Unknown");
+		b.append("|");
+		b.append(this.getActivePowerChannel().value().asString());
+
+		return b.toString();
+	}
+
+	@Override
+	public void handleEvent(Event event) {
+		if (!this.isEnabled()) {
+			return;
+		}
+
+		switch (event.getTopic()) {
+		case EdgeEventConstants.TOPIC_CYCLE_AFTER_PROCESS_IMAGE -> this.calculateEnergy();
+		case EdgeEventConstants.TOPIC_CYCLE_EXECUTE_WRITE -> {
+			this.executeWrite(this.getRelayChannel(), 0);
+		}
+		}
+	}
+
+	private void processHttpResult(JsonElement result, Throwable error) {
+		this._setSlaveCommunicationFailed(result == null);
+		if (error != null) {
+			this._setRelay(null);
```

Comment:
Out of experience I recommend to collect all values first and then set the Channel values all together. This way you will not forget to set a channel in any case. 

For example here you forget to (un)set `hasUpdate` and individual phase values in case of an error

Reply:
thanks !